camera_models.py

```diff
@@ -1129,4 +1130,171 @@ def get_rays_wrt_lidar_frame(self):
+        
+        NOTE: point_3d can also have a dimension of Bx3xN, where B is the 
+        batch number. 
+        
+        Returns: 
+        pixel_coor: A 2xN Tensor representing the 2D pixels. Bx2xN if batched.
+        valid_mask: A (N,) Tensor representing the valid mask. BXN if batched.
+        '''
+
+        point_3d = self.in_wrap(point_3d)
+
+        # torch.split results in Bx1XN.
+        x, y, z = torch.split( point_3d, 1, dim=-2 )
+
+        x2 = x**2.0 # Note: this may promote x2 to torch.float64 if point_3d.dtype=torch.int. 
+        y2 = y**2.0
+        z2 = z**2.0
+
+        # Magnitude of the 3D point.
+        R = (x2 + y2 + z2)**0.5
+
+        # Theta (th) is the angle between the ray and the z-axis.
+        # Phi (ph) is the angle between the ray and the x-axis, in the x-y plane.
+        ph = torch.acos( z / R )
+        th = torch.atan2( y, x )
+
+
+        # The radius in the image plane.
+        r = ph * self.ss.W / self.fov_rad
+
+        # The pixel coordinates.
+        # If any are FTensors, then convert to Tensor.
+        if isinstance(th, FTensor):
+            th = th.tensor()
+            ph = ph.tensor()
+
+
+        u = r * torch.cos(th) + self.ss.W / 2
+        v = r * torch.sin(th) + self.ss.H / 2
+
+        # Mask out the invalid pixels.
+        valid_mask = (u >= 0) & (u < self.ss.W) & (v >= 0) & (v < self.ss.H)
+        
+        # Pixel coordinates.
+        if normalized:
+            u = u / self.ss.W * 2 - 1
+            v = v / self.ss.H * 2 - 1
+
+        pixel_coor = torch.cat( (u, v), dim=-2 )
+
+        # Filter by FOV.
+        a = x2y2z_2_z_angle( x2, y2, z )
+        valid_mask = torch.logical_and(
+            valid_mask, 
+            a <= self.fov_rad / 2.0
+        )
+        
+        # This is for the batched dimension.
+        valid_mask = valid_mask.squeeze(-2)
+
+        return self.out_wrap(pixel_coor), self.out_wrap(valid_mask)
+    
+    def get_valid_bounary(self, n_points=1000):
+        '''
+        Get an array of pixel coordinates that represent the boundary of the valid region.
+        The result array is ordered.
+        '''
+        
+        # Unit length.
+        unit_length = torch.ones( (1, n_points), device=self.device, dtype=torch.float32 )
+        
+        # Find the x, y, z coordinates.
+        a_z = self.fov_rad / 2 # Angle w.r.t. the z-axis.
+        z = unit_length * math.cos( a_z )
+        
+        # Projection of the unit length onto the xy-plane.
+        r_xy = unit_length * math.sin( a_z )
+        a_x = torch.linspace( -LOCAL_PI, LOCAL_PI, n_points ) # Angle w.r.t. the x-axis.
+        x = r_xy * torch.cos( a_x )
+        y = r_xy * torch.sin( a_x )
+        
+        # Create an array of 3D points at the unit sphere along the FOV.
+        xyz = torch.cat( (x, y, z), dim=0 )
+        
+        pixel_coor, mask = self.point_3d_2_pixel(xyz)
+        return self.out_wrap(pixel_coor)
+    
+    def __str__(self) -> str:

```

Comment:
This needs to be revised. The member variables are wrong.

I am not so familiar with the `__repr__` and `__str__` magic functions. May be we need to revise all of them. I have the impression that `__repr__` should try to generate a string that can be used to re-generate the underlying object as python codes. `__str__` is some pretty string that friendly to humans for reading and understanding the object.

Reply:
Oh whoops yes I forgot about this one. Thank you for catching it!